### DISCUSSION TITLE
Set default type for each Model

### DISCUSSION THREAD

**Discussion by Mishanki** [spatie/laravel-tags#518] [Q&A]
This example create tag **without** type. Can I **set default type** for NewsItem model? 
```php
// create a model with some tags
$newsItem = NewsItem::create([
   'name' => 'testModel',
   'tags' => ['tag', 'tag2'], //tags will be created if they don't exist
]);
```

**Comment by Rocketpilot** [comment]
@Mishanki have you tried adding a 

`'type' => 'mydefaulttype'`

to the ::create array?

Edit: wait, I'm misremembering. Let me look at my code and I'll follow up.

**Comment by Rocketpilot** [comment]
Yeah I think the sample documentation is a bit misleading in terms of how you'd actually use the functionality in practice. If you're for instance creating a new blog entry in your BlogController, you'd have something like the following code inside your create post action:

```php
$blogPost->title = $request->title;
$blogPost->tag = Tag::findOrCreate($request->tag, 'myDefaultTagType');
```

You're essentially hard-coding the tag type in the controller.

**Reply by Mishanki** [reply]
I have type exception in latest laravel

```php
public function store(array $data): NewsItem
    {
        $data['tags'] = Tag::findOrCreate($data['tags'], 'news');

        return NewsItem::create($data);
    }
```

```php
NewsItem::setTagsAttribute(): Argument #1 ($tags) must be of type ArrayAccess|Spatie\Tags\Tag|array|string, null given,
called in .../vendor/laravel/framework/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php on line 1108
```

**Comment by Rocketpilot** [comment]
Hmm, try passing Tag::findOrCreate a string then, as it looks like it's being passed a null value.

```php
$data['tags'] = Tag::findOrCreate('this is the name of a tag', 'this-is-the-type-of-the-tag');
```

That's pretty well straight out of the docs.

**Reply by Mishanki** [reply]
Create tag with type  in `tag` table
After then throw exception when `NewsItem::create($data);`

Wrong:
```php
$data['tags'] = Tag::findOrCreate($data['tags'], 'news'); // <--- yes, it`s create tag with type in tag table
NewsItem::create($data); // <-- exception
```

In doc:
```php
//creating a tag with a certain type
$tagWithType = Tag::findOrCreate('headline', 'newsTag');
// attach/detach
$newsItem->attachTag($tagWithType);
$newsItem->detachTag($tagWithType);
// sync
$newsItem->syncTagsWithType(['tagA', 'tagB'], 'firstType');
$newsItem->syncTagsWithType(['tagC', 'tagD'], 'secondType');
```

**Reply by Rocketpilot** [reply]
Can you show me NewsItem.php, your model file?

**Reply by Mishanki** [reply]
Like this

```php
class SomeModel extends Model
{
    use HasFactory;
    use HasTags;
    use HasUlids;

    /** @var bool */
    public $incrementing = false;
    /** @var bool */
    public $timestamps = true;
    /** @var string */
    protected $primaryKey = 'id';
    /** @var string */
    protected $keyType = 'string';
    /** @var string */
    protected $table = 'some_model';
    /** @var bool */
    protected static $unguarded = true;
    /** @var array<string,string> */
    protected $casts = [
    ];
}
```

**Reply by Rocketpilot** [reply]
Are you able to successfully save instances of NewsItem to your database without attaching a tag to them?

**Reply by Mishanki** [reply]
Yes, I can save data to DB via model. This **native** way still work for me too. But I think **native way not work with type**. I want use type without coding (like native).

Expect way to use tags with type
```php
$newsItem = NewsItem::create([
   'name' => 'testModel',
   'tags' => [['tag', 'tag2'], 'some-type'], 
]);
```

**Reply by Rocketpilot** [reply]
To be honest I think this is a harder way to do things. I’ve been building a sample site to get to grips with this functionality and I wrote a TagController and front end to create the tags separately with types and then a form to add the existing tags to the Post model.

**Reply by Mishanki** [reply]
Thank you for help and your time

**Reply by Rocketpilot** [reply]
Good luck with it! Maybe someone with more experience under their belt than I have will comment. You could also ask on the Laravel subreddit as it’s pretty popular.
